server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,217 @@
+# Proposal: Update Workflow Server Interface
+
+Currently (as of version 1.15), ad-hoc runtime interaction with Temporal
+workflows is achieved through queries and signals. The former technique can
+interrogate the state of a workflow and returns a result but must not _change_
+the state of the workflow, while the latter can change workflow state but does
+not natively support return values [^1]. 
+
+The missing piece in runtime workflow interaction is an affordance that
+provides both (1) the ability to change workflow state, and (2) a mechanism to
+observe the result of the same without polling or setting up callback
+structures. This feature will be workflow updates.
+
+This proposal describes the server API (i.e. gRPC) for such a feature. We leave
+implementation details for a deeper design review and instead relate the API,
+its expected usage and the guarantees provided.
+
+## Concepts
+
+### Update
+
+An Update is the execution of arbitrary workflow logic (subject to the standard
+workflow code constraints on determinism) triggered by an UpdateRequest and
+eventually resulting in a single UpdateResponse. The processing of an Update is
+performed by workflow code and as such the timeliness of the UpdateResponse is
+dependent on worker resources being available to process the Update. A given
+Update applies to exactly one workflow execution. The lifetime of an updated
+begins at the first UpdateRequest triggering said Update and ends when the
+UpdateResponse has been made durable. Consequently, the lifetime of an Update
+may exceed the lifetime of the RPC request/response that initiates it.
+
+### Update ID
+
+The unique identifier of an Update. Generated by the Temporal server.
+
+### UpdateRequest
+
+An UpdateRequest is a remote call originating either from outside the Temporal
+system entirely or from another Temporal workflow.  UpdateRequests are
+deduplicated such that issuance of the same UpdateRequest multiple times is
+effectively the same as issuing a single request. The UpdateRequest includes an
+indication of the workflow execution at which the Update is targeted.
+UpdateRequests are issued over gRPC or through SDK-specific APIs wrapping the
+same. Both issuing and receiving UpdateRequests becomes a durable part of a
+given workflow's history.
```

Comment:
So obviously I can re-issue a known update to get what the response _was_. Some questions here:

1. Does that work even after the workflow is complete? EDIT: I see below that they may be accessed after workflow complete, though unsure if we'll provide an easy way.
2. Is there anyway to say "only-if-already-issued" so the update isn't an upsert of sorts, but really just a get-result-if-update-present? Maybe there's no use case for this?

Reply:
The combination of an update ID always dereferencing to _some_ result _eventually_ and invalid UpdateRequests not causing history traffic means that it _has_ to be the server generating the update ID. If you want something like "do X if you haven't already done it" the client could use a hash of the UpdateRequest as the request_id.

Reply:
Request ID's are traditionally auto-generated by our SDK for use by the server for their idempotency reasons of the literal gRPC request. Not to tie potentially separate requests together. Are we expecting the user to provide this request ID? I wonder if there is a better name (I can't think of one off-hand).

Reply:
> There is no user-facing idempotence

There is though? Issue two UpdateReqests with the same request_id and they get coalesced/deduped